Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,687 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import GRPCCore
+import NIOCore
+
+enum OnMetadataReceived {
+  case reject(status: Status, trailers: Metadata)
+  case doNothing
+}
+
+fileprivate protocol GRPCStreamStateMachineProtocol {
+  var state: GRPCStreamStateMachineState { get set }
+  
+  mutating func send(metadata: Metadata) throws
+  mutating func send(message: [UInt8], endStream: Bool) throws
+  mutating func send(status: String, trailingMetadata: Metadata) throws
+  
+  mutating func receive(metadata: Metadata, endStream: Bool) throws -> OnMetadataReceived
+  mutating func receive(message: ByteBuffer, endStream: Bool) throws
+  
+  mutating func nextOutboundMessage() throws -> ByteBuffer?
+  mutating func nextInboundMessage() -> [UInt8]?
+}
+
+enum GRPCStreamStateMachineConfiguration {
+  case client(maximumPayloadSize: Int)
+  case server(
+    maximumPayloadSize: Int,
+    supportedCompressionAlgorithms: [Encoding]
+  )
+}
+
+enum GRPCStreamStateMachineState {
+  case clientIdleServerIdle(ClientIdleServerIdleState)
+  case clientOpenServerIdle(ClientOpenServerIdleState)
+  case clientOpenServerOpen(ClientOpenServerOpenState)
+  case clientOpenServerClosed(ClientOpenServerClosedState)
+  case clientClosedServerIdle(ClientClosedServerIdleState)
+  case clientClosedServerOpen(ClientClosedServerOpenState)
+  case clientClosedServerClosed
+
+  struct ClientIdleServerIdleState {
+    let maximumPayloadSize: Int
+  }
+  
+  struct ClientOpenServerIdleState {
+    var framer: GRPCMessageFramer
+    var compressor: Zlib.Compressor?
+    
+    let deframer: NIOSingleStepByteToMessageProcessor<GRPCMessageDeframer>
+    var decompressor: Zlib.Decompressor?
+    
+    var inboundMessageBuffer: OneOrManyQueue<[UInt8]>
+    
+    init(
+      previousState: ClientIdleServerIdleState,
+      compressionAlgorithm: Encoding?
+    ) {
+      if let zlibMethod = Zlib.Method(encoding: compressionAlgorithm) {
+        self.compressor = Zlib.Compressor(method: zlibMethod)
+        self.decompressor = Zlib.Decompressor(method: zlibMethod)
+      }
```

Comment:
These can actually be different: the client e.g. set "message-encoding" to "deflate" and "message-accept-encoding" to "identity, gzip" meaning that they will send compressed messages with accept responses which are either not compressed or compressed with "gzip"